File: primalTestDotNet/AI/States/LookingForPotionState.cs
using PrimalTestDotNet.Entities;

namespace PrimalTestDotNet.AI.States;

public class LookingForPotionState(AIStateMachine context, Level.Level level) : StateBase(level), IState
{
    public void Update(Hero hero, IEnumerable<IGameObject> gameObjects)
    {
        if (hero.Health > 1)
        {
            context.ChangeState(context.LookingForTreasureState);
            return;
        }

        // Look for potion
        var allMonsters = gameObjects.Where(o => o is Monster);
        var recalculatedDistanceData = CalculateDistanceData(hero.Position, allMonsters);
        var potion = gameObjects.Where(o => o is Potion).FirstOrDefault();
        if (potion == null)
        {
            context.ChangeState(context.UnwinnableState);
            return;
        }
        var pathToPotion = GetPath(hero.Position, potion.Position, recalculatedDistanceData);

        var node = pathToPotion.Last();
        var direction = node - hero.Position;
        hero.Move(direction);
    }
}

File: PrimalTestCpp/LookingForPotionState.cpp
#include "LookingForPotionState.h"
#include "Hero.h"
#include "AIStateMachine.h"
#include "Monster.h"
#include "Potion.h"

LookingForPotionState::LookingForPotionState(AIStateMachine& context, Level& level)
	: StateBase(level), _context(context) { }

void LookingForPotionState::update(Hero& hero, std::vector<std::shared_ptr<IGameObject>>& gameObjects){
	if (hero.getHealth() > 1) {
		_context.changeState(_context.LOOKING_FOR_TREASURE_STATE());
		return;
	}

	auto potions = findAllObjectsOfType<Potion>(gameObjects);
	if (potions.size() == 0) {
		_context.changeState(_context.UNWINNABLE_STATE());
		return;
	}

	auto allMonsters = findAllObjectsOfType<Monster>(gameObjects);
	// monsters are walls...
	auto recalculatedDistanceData = calculateDistanceData(hero.getPosition(), allMonsters);
	auto& potion = potions.front();
	auto pathToPotion = getPath(hero.getPosition(), potion->getPosition(), recalculatedDistanceData);

	auto& node = pathToPotion.back();
	auto direction = node - hero.getPosition();

	// No clear path to the potion
	if (direction.getLength() > 1) {
		_context.changeState(_context.UNWINNABLE_STATE());
		return;
	}

	hero.move(direction);
}
